Feature/ts3 adapter

Jonsen94 opened the pull request:
Adds basic ts3 functionalities.
- connects to a ts3 server
- rest call to get a client list by ip
- send password to user in ts3

Commit:
ignoring /bin/ (eclipse)
added ts3 manager to handle the ts3api, query and config
clientDbId is Integer now
Added basics of the ts3Login Module

Commit:
bot has a nickname now and chooses virtual server after logging in

Commit:
enhanced gitignore

Commit:
better log levels
some javadoc todos
added message builder

Commit:
Merge branch 'master' into feature/ts3Adapter

# Conflicts:
#	.gitignore
#	src/main/java/eu/kandru/luna/controller/AuthController.java

Commit:
fixed nickname
added rest request for userlist by ip
better test login messages

Commit:
added a test for the TS3LoginModule class

Commit:
javadoc for the teamspeak classes

ja-ko commented on src/main/java/eu/kandru/luna/controller/AuthController.java:
```diff
@@ -32,18 +50,41 @@
+			}
+		} catch (TimeoutException e) {
+			log.warn("Receiving client list from ip timed out.");
+			response.setStatus(500);
```
Replace with HttpServletResponse.SC_something

ja-ko commented on src/main/java/eu/kandru/luna/controller/AuthController.java:
```diff
@@ -57,6 +98,7 @@ public AuthChallengeResponse challenge(@RequestBody AuthChallengeRequest request
                                .build()
                                .toJwt(jwtService);
         log.info("generated password " + password);
+		ts3LoginModule.sendPasswordToUser(request.getClientDbId(), password);
```
fix indentation

ja-ko commented on src/main/java/eu/kandru/luna/teamspeak/TS3Manager.java:
```diff
@@ -0,0 +1,71 @@
+
+			CommandFuture<Boolean> loginCommand = ts3api.login(ts3props.getLogin(), ts3props.getPassword());
+			loginCommand.onSuccess(result -> {
+				log.info("Connected to ts3 server.");
```
We should consider logging some part of the result. It miht have valueable information (especially in the error case below)

ja-ko commented on src/main/java/eu/kandru/luna/teamspeak/TS3Manager.java:
```diff
@@ -0,0 +1,71 @@
+				log.info("Connected to ts3 server.");
+			});
+			loginCommand.onFailure(result -> {
+				log.error("Failed to connect to ts3 server.");
```
see above

ja-ko commented on src/main/java/eu/kandru/luna/teamspeak/TS3Manager.java:
```diff
@@ -0,0 +1,71 @@
+			
+			CommandFuture<Boolean> registerCommand = ts3api.registerAllEvents();
+			registerCommand.onFailure(result -> {
+				log.error("Couldnt register all ts3 events.");
```
see above

ja-ko commented on src/main/java/eu/kandru/luna/teamspeak/modules/login/TS3LoginModule.java:
```diff
@@ -0,0 +1,105 @@
+@Component
+public class TS3LoginModule {
+	private TS3ApiAsync ts3api;
+	private final int REQUEST_TIMEOUT = 10000;
```
Make this configurable for all ts3 queries

ja-ko commented on src/main/resources/application.yml:
```diff
@@ -23,3 +23,4 @@ teamspeak:
     login:
     password:
     server-id:
+    nickname: Luna
```
Add a newline at the end of the file

ja-ko commented on src/test/java/eu/kandru/luna/controller/AuthControllerTest.java:
```diff
@@ -36,6 +42,9 @@
 
```
Please use this test class to assert that a password is actually sent to the correct client

Commit:
Status code now with enums
Better log messages when conencting to the ts server and when registering events

Commit:
added newline at the end of the file

Commit:
small fixes for the log messages

Commit:
replaced tab indentation with space indentation

Commit:
password is now in bold
request timeout is a configurable ts3propertie
login module uses the ts3properties
more default config in the application.yml
fixed test

Commit:
Created a test for the send password method from the TS3LoginModule and small fixes
-TS3LoginModule now uses better logging pattern
-TS3LoginModule tests now contains a test for the send password method
-Replaced tabs with spaces for the test class

ja-ko commented on src/test/java/eu/kandru/luna/teamspeak/modules/login/TS3LoginModuleTest.java:
```diff
@@ -24,47 +30,92 @@
+        verify(ts3api).sendPrivateMessage(eq(42), messageCaptor.capture());
+        
+        if(!messageCaptor.getValue().contains(password)){
+            fail("Password message did not contain password. Expected password \""+password+"\", received message \""+messageCaptor.getValue()+"\".");
```
This is why god made asserts. Please use them.

ja-ko reviewed (changes requested):
[AuthControllerTest::authenticateSuccessful()](https://github.com/Kandru/ts3luna/blob/master/src/test/java/eu/kandru/luna/controller/AuthControllerTest.java#L85) could be changed use the actually provided password by mocking the underlying ts3api.
However, I'd accept the PR as soon as you switched the fail for a proper assert.

Commit:
Enhanced login module tests, better mocking for the ts3api
- Added abstract class which takes care of mocking the ts3api for test classes
- Changed the login module tests so that they use the new ts3api mocking class

Commit:
Finished and fixed tests, some small api tweaks
-autowiring ts3apiasync direct now instead of inderect through the ts3manager
-AuthControllerTest now tests if the right password was send to the client
-AbstractTS3ApiMock now uses MockBean instead of Mock
-fixed AbstractTS3ApiMock setup method
-fixed TS3LoginModuleTest

Commit:
Password is now extracted from the password message during tests
OneTimePasswordGenerator isn't mocked anymore

Commit:
added test to check if the OneTimePasswordGenerator generates different passwords

ja-ko commented on src/test/java/eu/kandru/luna/controller/AuthControllerTest.java:
```diff
@@ -110,9 +107,13 @@ private ResultActions authenticate(String challenge, String password) throws Exc
-        MvcResult result = authenticate(challenge, PASSWORD).andExpect(status().isOk())
+    	verify(ts3api).sendPrivateMessage(eq(CLIENT_ID), passwordTextCaptor.capture());    	
+    	Pattern p = Pattern.compile("[0-9]{6}");
+        Matcher m = p.matcher(passwordTextCaptor.getValue());
```
fix indentation

ja-ko commented on src/test/java/eu/kandru/luna/util/OneTimePasswordGeneratorTest.java:
```diff
@@ -0,0 +1,33 @@
+		for (int i = 0; i < PASSWORDS_TO_CHECK; i++){
+			passwords[i] = passwordGenerator.generatePassword();
+		}
+		for (int i = 0; i < PASSWORDS_TO_CHECK-1; i++){
```
we gotta talk about this one :)

Commit:
-fixed indent
-better password generator test
